src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningConditionMapper.java

```diff
@@ -0,0 +1,367 @@
+            this.valuePlaceholders = valuePlaceholders;
+        }
+
+        @Override
+        public Void visitKeyConditionPart(KeyConditionPartContext keyConditionPart) {
+            IdContext id = keyConditionPart.id();
+            if (id != null) {
+                String fieldName = keyConditionPart.id().getText();
+                if (fieldName.startsWith("#")) {
+                    // if it's a field placeholder, remove it -- we'll generated a new one for translated expression
+                    fieldName = removeFieldPlaceholder(fieldName);
+                }
+
+                if (primaryKey.getHashKey().equals(fieldName)) {
+                    checkArgument(hkValue == null, "Multiple hash key conditions");
+                    ComparatorContext comparator = keyConditionPart.comparator();
+                    checkArgument(comparator != null && comparator.getText().equals("="),
+                        "Hash key condition must have operator '='");
+                    String valuePlaceholder = keyConditionPart.literal(0).getText();
+                    hkValue = removeValuePlaceholder(valuePlaceholder);
+                } else if (primaryKey.getRangeKey().isPresent() && primaryKey.getRangeKey().get().equals(fieldName)) {
+                    checkArgument(rkCondition.isEmpty(), "Multiple range key conditions");
+                    ComparatorContext comparator = keyConditionPart.comparator();
+                    ComparisonOperator op = comparator != null ? getComparatorType(comparator.getText()) : BETWEEN;
+                    List<AttributeValue> values = keyConditionPart.literal().stream()
+                        .map(LiteralContext::getText)
+                        .map(this::removeValuePlaceholder)
+                        .collect(Collectors.toList());
+                    rkCondition = Optional.of(new KeyFieldCondition(op, values));
+                } else {
+                    throw new IllegalArgumentException("Key condition expression contains non-key field: " + fieldName);
+                }
+            }
+            return visitChildren(keyConditionPart);
+        }
+
+        private AttributeValue removeValuePlaceholder(String placeholder) {
+            AttributeValue value = null;
+            if (valuePlaceholders != null) {
+                value = valuePlaceholders.remove(placeholder);
+            }
+            checkArgument(value != null, "Referenced attribute value does not exist: %s", placeholder);
+            return value;
+        }
+
+        private String removeFieldPlaceholder(String placeholder) {
+            String literal = null;
+            if (fieldPlaceholders != null) {
+                literal = fieldPlaceholders.remove(placeholder);
+            }
+            checkArgument(literal != null, "Referenced attribute name does not exist: %s", placeholder);
+            return literal;
+        }
+    }
+
+    private static ComparisonOperator getComparatorType(String comparator) {
```

Comment:
Antlr doesn't do this for you?

Reply:
The return value here is a com.amazonaws.services.dynamodbv2.model.ComparisonOperator